### ISSUE TITLE
Argument "key" to "sorted" has incompatible type "Callable…"

### ISSUE THREAD

**Issue by roubert** [python/mypy#19259]
**Bug Report**

I believe that I've found a bug in mypy, the following MRE should as far as I can tell not result in any typing errors at all, but it reports an error (and I've found two simple workarounds for which no error is reported, strengthening my belief that this is a bug and not intended behaviour).

**To Reproduce**

```python
#!/usr/bin/python3

from collections.abc import Iterable
from typing import cast


def unproblematic(iter: Iterable[int]) -> None:
    print(iter)


def problematic(iter: Iterable[int | str]) -> None:
    print(iter)


def key(item: int) -> int:
    return -item


def main() -> None:
    items: list[int] = [3, 4, 1, 2]

    unproblematic(sorted(items, key=key))  # Works.
    problematic(sorted(items, key=key))  # Fails mypy arg-type.

    workaround = sorted(items, key=key)
    problematic(workaround)  # Works.

    problematic(cast(Iterable[int], sorted(items, key=key)))  # Works.


if __name__ == '__main__':
    main()
```

**Actual Behavior**

```sh
$ mypy mre.py
mre.py:23: error: Argument "key" to "sorted" has incompatible type "Callable[[int], int]"; expected "Callable[[int | str], SupportsDunderLT[Any] | SupportsDunderGT[Any]]"  [arg-type]
```

**Your Environment**

```sh
$ mypy --version                
mypy 1.16.0 (compiled: yes)
$ python3 --version
Python 3.13.3
```

(If anyone can't trivially reproduce this, I'd be happy to provide more details about my environment.)

**Comment by sterliakov** [comment]
That's yet another case where applying return type context and args context separately causes overly broad inference.
